### PR TITLE
Proof of concept of adding options per-requirement

### PR DESCRIPTION
**Pull request by memsharded**
Responding to https://github.com/conan-io/conan/pull/9839, and to the need of defining different options in build_requires and requires:

- This approach can actually set any option for any type of requirement, at the individual level (can work for multiple regular requires of different version of same package (if no conflicts))

Close https://github.com/conan-io/conan/pull/9839

### CODE REVIEW

**Review by franramirez688 — APPROVED** [review]

File: conans/test/integration/options/test_options_build_requires.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    client.save({"openssl_1_1_1.py": conanfile_openssl_1_1_1,
+                 "openssl_3_0_0.py": conanfile_openssl_3_0_0,
+                 "conanfile_cmake.py": conanfile_cmake,
+                 "conanfile.py": conanfile_consumer})
+
+    client.run("create openssl_1_1_1.py")
+    client.run("create openssl_3_0_0.py")
+    client.run("create conanfile_cmake.py")
+    client.run("install conanfile.py")
```
Comment (lasote) [inline conans/test/integration/options/test_options_build_requires.py:32]: so?

**Review by lasote — APPROVED** [review]
I know these tests are pretty straightforward but we must introduce the dev rule of a docstring on every test. That would save a lot of review time and debug time when a feature breaks some test.

**Comment by memsharded** [comment]
Agree with the docs on tests. This is the review I did to other PR after @franramirez688 suggestion: https://github.com/conan-io/conan/pull/10136/commits/5ec4b5a95c923d83fc90454f8ddda23edf2792f5. Lets do it always.

File: conans/test/integration/options/test_options_build_requires.py (hunk 1 above)
Comment (memsharded) [reply conans/test/integration/options/test_options_build_requires.py:32]: Added checks and some explanations, as well as docstrings for tests

**Comment by memsharded** [comment]
Missing @czoido, but the idea seems to be ok? I submitted this as a PoC for the interface, I will try to extend a bit more the tests and mark it ready for review & merge

**Review by czoido — APPROVED** [review]

**Comment by lasote** [comment]
OMG so happy to see these docstrings

File: conans/client/graph/graph_builder.py [hunk 2]
```diff
@@ -250,8 +250,18 @@ def _create_new_node(self, node, require, graph, profile_host, profile_build, gr
...
-        # FIXME
-        down_options = node.conanfile.up_options
+        # The consumer "up_options" are the options that come from downstream to this node
+        if require.options is not None:
+            # If the consumer has specified requires(options=xxx), we need to use it
+            # It will have less priority than downstream consumers
+            down_options = Options(options_values=require.options)
+            down_options.scope(new_ref.name)
+            # TODO: discuss, together with build_require override, if it is "build" or "visible"
+            if require.visible:  # Only visible requirements propagate options from downstream
```
Comment (memsharded) [inline conans/client/graph/graph_builder.py]: Changed this from ``not require.build`` => ``require.visible``., thinking about what makes more sense...

**Comment by memsharded** [comment]
I have added tests to validate that ``visible=False`` requires cannot be overriden/forced from downstream.
Also including the options propagation condition to not apply to visible=False and to apply only to context=HOST.

Please re-review.

File: conans/client/graph/graph_builder.py [hunk 3]
```diff
@@ -250,10 +250,19 @@ def _create_new_node(self, node, require, graph, profile_host, profile_build, gr
...
+        # The consumer "up_options" are the options that come from downstream to this node
         if require.options is not None:
-            down_options.update_options(Options(options_values=require.options))
+            # If the consumer has specified requires(options=xxx), we need to use it
+            # It will have less priority than downstream consumers
+            down_options = Options(options_values=require.options)
+            down_options.scope(new_ref.name)
+            # TODO: discuss, together with build_require override, if it is "build" or "visible"
+            if require.visible and context == CONTEXT_HOST:
+                # Only visible requirements in the host context propagate options from downstream
```
Comment (lasote) [inline conans/client/graph/graph_builder.py:264]: I would add a comment to recall the discussion about the `build_default_options` or whatever name and that we decided to be more restrictive at this point until we have stronger arguments to decide what to do.

**Review by lasote — APPROVED** [review]

**Review by czoido — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]

**Pull request merged**